roaring/roaring.go

```diff
@@ -395,32 +399,282 @@ func (b *Bitmap) Intersect(other *Bitmap) *Bitmap {
+// in place.
+//
+// This function performs an n-way union of n bitmaps. It performs this in an
+// optimized manner looping through all the bitmaps and performing unions one
+// container at a time. As a result, instead of generating many intermediary
+// containers for each union operation for a given container key, only one
+// new container needs to be allocated (or re-used) regardless of how many bitmaps
+// participate in the union. This significantly reduces allocations. In addition,
+// because we perform the unions one container at a time across all the bitmaps, we
+// can calculate summary statistics that allow us to make more efficient decisions
+// up front. For example, imagine trying to perform a union across the following three
+// bitsets:
+//
+//     1. Bitmap A: Single array container at key 0 with 400 values in it.
+//     2. Bitmap B: Single array container at key 0 with 500 values in it.
+//     3. Bitmap C: Single array container at key 0 with 3500 values in it.
+//
+// Naive approach:
+//
+//     1. Perform union of bitmap A and B, container by container
+//         a. 400 + 500 < ArrayMaxSize so likely we will choose to allocate a new array
+//            container and then perform a unionArrayArray operation to merge the two
+//            arrays into the new array container.
+//     2. Perform a union of the bitmap generated in the step above with bitmap C.
+//        900 + 3500 > ArrayMaxSize so we will need to upgrade to a bitset container which
+//        we will have to allocate, and then we will have to perform two unions into the
+//        new bitmap container: one for the array container generated in the previous step,
+//        and one for the bitset container in bitmap C.
+//
+// Approach taken by this function:
+//
+//     1. Detect that bitmaps A, B, and C all have containers for key 0.
+//     2. Estimate the resulting cardinality of the union of all their containers to be
+//        400 + 500 + 3500 > ArrayMaxSize and decide upfront to use a bitset for the target
+//        container.
+//     3. Union the containers from bitmaps A, B, and C into the new bitset container directly
+//        using fast bitwise operations.
+//
+// In the naive approach, we had to allocate two containers, whereas in the optimized approach
+// we only had to allocate one container, and we also had to perform less union operations. This
+// example is simplistic, but the impact in terms of CPU cycles and memory allocations achieved
+// by using the optimized alogorithm when unioning many large bitmaps can be huge.
+//
+// An additional optimization that this function makes is that it recognizes that even when
+// CPU support is present, performing the popcount() operation isn't free. Imagine a scenario
+// where 10 bitset containers are being unioned together one after the next. If every
+// bitset<->bitset union operation needs to keep the containers cardinality up to date, then
```

Comment:
I think `containers'`, because that's a possessive.

Reply:
sure thing

Reply:
fixed!